Replace libLLVM symlink with linker script

nikic opened the pull request:
It turns out that the libLLVM-N.so -> libLLVM.so.N.1 symlink is also needed when projects like miri link against librustc_driver.so. As such, we have to distribute it in real rustup components like rustc-dev, rather than only for download-ci-llvm.

To avoid actually distributing symlinks (which are not supported or not fully supported by the rustup infrastructure) replace it with a linker script that does the same thing instead.

Fixes https://github.com/rust-lang/rust/issues/121889.

r? @cuviper

nikic commented:
@bors try

bors commented:
:hourglass: Trying commit 3501f4aea596ecbc456c1c38751fc791fe3c866e with merge f73e623e5edfb62d8c85784c0c5acb62ea094813...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3501f4aea596ecbc456c1c38751fc791fe3c866e","merge_sha":"f73e623e5edfb62d8c85784c0c5acb62ea094813"} -->

bjorn3 commented on src/bootstrap/src/core/build_steps/dist.rs:
```diff
@@ -2035,11 +2030,15 @@ fn install_llvm_file(
+        // Replace the symlink itself with an equivalent linker script. This is used by
+        // download-ci-llvm and projects linking against librustc_driver.so. We don't use a
+        // symlink, as these are not allowed inside rustup components.
+        let link = t!(fs::read_link(source));
```
This will be a relative path, right?

nikic replied:
Yes, this will produce something like `INPUT(libLLVM.so.18.1-rust-dev)`.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8137560357/job/22239350784)
Build commit: f73e623e5edfb62d8c85784c0c5acb62ea094813 (`f73e623e5edfb62d8c85784c0c5acb62ea094813`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8137560357/job/22239350784"},"merge_sha":"f73e623e5edfb62d8c85784c0c5acb62ea094813"} -->

nikic commented:
Breaks download-ci-llvm, because that doesn't have the symlink now either, so the new logic doesn't apply to it anymore.

I'll restore the symlink for the download-ci-llvm case.

nikic committed:
Replace libLLVM symlink with linker script in rustup components

It turns out that the libLLVM-N.so -> libLLVM.so.N.1 symlink is
also needed when projects like miri link against librustc_driver.so.
As such, we have to distribute it in real rustup components like
rustc-dev, rather than only for download-ci-llvm.

To avoid actually distributing symlinks (which are not supported
or not fully supported by the rustup infrastructure) replace it
with a linker script that does the same thing instead.

nikic commented:
@bors try

bors commented:
:hourglass: Trying commit 5d1d4081d9e5b2d8748bd3f7b7403a5dffee69b1 with merge 618690d791b4cc1d90b113179a25ae3cd7189ed2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5d1d4081d9e5b2d8748bd3f7b7403a5dffee69b1","merge_sha":"618690d791b4cc1d90b113179a25ae3cd7189ed2"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8138618604/job/22243202757)
Build commit: 618690d791b4cc1d90b113179a25ae3cd7189ed2 (`618690d791b4cc1d90b113179a25ae3cd7189ed2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8138618604/job/22243202757"},"merge_sha":"618690d791b4cc1d90b113179a25ae3cd7189ed2"} -->

nbdd0121 reviewed (approved):
Code looks good, although it's a bit sad that w have separate code path for download-ci-llvm and everything else.

nikic commented:
I've confirmed that download-ci-llvm works again now, and also confirmed that miri can be built against this toolchain (I had to edit miri-script to disable rustfmt and clippy, which are not part of the try toolchain).

nikic commented:
> > Breaks download-ci-llvm, because that doesn't have the symlink now either, so the new logic doesn't apply to it anymore.
> 
> Why does the linker script not work for download-ci-llvm?

It does work for the purposes of linking, but it breaks the code being modified here that handles libLLVM installation. The logic is based around having the symlink. If we don't ship the symlink for download-ci-llvm, then the logic would additionally also have to support reading the linker script to determine which shared object it needs to install in the libdir.

RalfJung commented:
Oh, this function later works on its own output as input? Wow that's recursive.^^

Might be worth having a comment in the code explaining that.

bjorn3 commented:
On Windows 10 it still requires developer mode to be enabled, right?

mati865 commented:
Both Windows 10 and 11 require either developer mode or modification of group policies.

RalfJung commented:
@cuviper this is blocking all tools linking against rustc_driver, would be nice to get it landed soon. :)

Mark-Simulacrum commented:
@bors r+

The risk here seems fairly low (I expect to hit some problems but that's almost always true with changes like this). In the meantime, it should unblock miri etc.

RalfJung commented:
@bors p=1

bors commented:
:hourglass: Testing commit 5d1d4081d9e5b2d8748bd3f7b7403a5dffee69b1 with merge bfe762e0ed2e95041cc12c02c5565c4368f2cc9f...
<!-- homu: {"type":"BuildStarted","head_sha":"5d1d4081d9e5b2d8748bd3f7b7403a5dffee69b1","merge_sha":"bfe762e0ed2e95041cc12c02c5565c4368f2cc9f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8174036196/job/22354902057)
Approved by: Mark-Simulacrum
Pushing bfe762e0ed2e95041cc12c02c5565c4368f2cc9f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8174036196/job/22354902057"},"merge_sha":"bfe762e0ed2e95041cc12c02c5565c4368f2cc9f"} -->

rust-timer commented:
Finished benchmarking commit (bfe762e0ed2e95041cc12c02c5565c4368f2cc9f): [comparison URL](https://perf.rust-lang.org/compare.html?start=3314d5ce4c209e840c2e4b2c4442f6e031ae0989&end=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3314d5ce4c209e840c2e4b2c4442f6e031ae0989&end=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3314d5ce4c209e840c2e4b2c4442f6e031ae0989&end=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.0%, 4.2%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [0.9%, 3.6%]   | 12    |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.4%, -1.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.7%  | [-1.4%, 4.2%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3314d5ce4c209e840c2e4b2c4442f6e031ae0989&end=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=3314d5ce4c209e840c2e4b2c4442f6e031ae0989&end=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 646.923s -> 644.65s (-0.35%)
**Artifact size**: 175.06 MiB -> 175.06 MiB (0.00%)
